### PR TITLE
WIP: use low level sixel_encode

### PR DESCRIPTION
**Pull request by johnnychen94**
MWE is in `demo.jl`:

```julia
using Sixel, ImageCore, TestImages, ImageInTerminal
img = testimage("cameraman")

enc = Sixel.SixelEncoder(stdout, img)
enc(img)
```

cc: @Gnimuc

### CODE REVIEW

File: src/encoder.jl [hunk 1]
```diff
@@ -13,18 +13,33 @@ struct SixelEncoder{I<:IO} <: AbstractEncoder
...
     SixelEncoder(io, colorbits, pixelformat, allocator)
 end
 
 function sixel_write_callback_function(buffer_ptr::Ptr{Cchar}, sz::Cint, priv::Ref{T})::Cint where {T<:IO}
     io = unsafe_load(Base.unsafe_convert(Ptr{T}, priv))
     buffer = unsafe_wrap(Array{Cchar}, buffer_ptr, (sz, ))
-    write(io, buffer)
+
+    maybe_restore!(io)
+    outsz = write(io, buffer)
```
Comment (Gnimuc) [inline src/encoder.jl:39]: `unsafe_write(io, buffer_ptr, sz)`

File: src/encoder.jl [hunk 2]
```diff
@@ -13,18 +13,33 @@ struct SixelEncoder{I<:IO} <: AbstractEncoder
...
 
     colorbits = default_colorbits(img)
     pixelformat = default_pixelformat(img)
+    io = maybe_convert(SixelIOBuffer, io)
     SixelEncoder(io, colorbits, pixelformat, allocator)
 end
 
 function sixel_write_callback_function(buffer_ptr::Ptr{Cchar}, sz::Cint, priv::Ref{T})::Cint where {T<:IO}
     io = unsafe_load(Base.unsafe_convert(Ptr{T}, priv))
     buffer = unsafe_wrap(Array{Cchar}, buffer_ptr, (sz, ))
```
Comment (Gnimuc) [inline src/encoder.jl:36]: `# buffer = unsafe_wrap(Array{Cchar}, buffer_ptr, (sz, ))`

File: src/encoder.jl [hunk 3]
```diff
@@ -13,18 +13,33 @@ struct SixelEncoder{I<:IO} <: AbstractEncoder
...
     allocator = SixelAllocator()
 
     colorbits = default_colorbits(img)
     pixelformat = default_pixelformat(img)
+    io = maybe_convert(SixelIOBuffer, io)
     SixelEncoder(io, colorbits, pixelformat, allocator)
 end
 
 function sixel_write_callback_function(buffer_ptr::Ptr{Cchar}, sz::Cint, priv::Ref{T})::Cint where {T<:IO}
     io = unsafe_load(Base.unsafe_convert(Ptr{T}, priv))
```
Comment (Gnimuc) [inline src/encoder.jl:35]: Oops! This `io` is created by `unsafe_load` which is a copy of the value of the priv(a `Ref`).

In fact, there is no need to do this `unsafe_load`. `priv` is already a `Ref{IO}` type as we use the parametric type. So the body of this callback function should be:
 
```
unsafe_write(priv[], buffer_ptr, sz)
```

That's it. I'm sorry for the confusion.

**Comment by johnnychen94** [comment]
Since the low-level interop calling issue is now fixed, I'm merging this into the main PR #1. Thank you @Gnimuc for the helps!

**Pull request merged**